llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -25173,6 +25173,37 @@ static SDValue combineConcatVectorOfCasts(SDNode *N, SelectionDAG &DAG) {
   return DAG.getNode(CastOpcode, DL, VT, NewConcat);
 }
 
+static SDValue combineConcatVectorInterleave(SDNode *N, SelectionDAG &DAG) {
+  SDValue FirstOp = N->getOperand(0);
+  if (FirstOp.getOpcode() != ISD::VECTOR_INTERLEAVE ||
+      FirstOp.getNumOperands() != N->getNumOperands())
+    return SDValue();
+
+  for (unsigned I = 0; I < N->getNumOperands(); I++) {
```

Comment:
Nit, does enumerating op_values also work here?
```suggestion
  for (auto [Idx, Op] : enumerate(N->op_values())) {
```